File: Kotlin/LambdaCalculus.kt
/**
 * Lambda Calculus Implementation in Kotlin
 * Created by Claude (Anthropic, PBC), version 3.5 Sonnet
 * 
 * Recommended filename: LambdaCalculus.kt
 * 
 * Usage: This file can be run in a Kotlin REPL or compiled and executed
 * as a Kotlin program. Each section demonstrates different concepts of
 * Lambda Calculus implemented in Kotlin.
 */

// Step 1: Understand Lambda Expressions
typealias LambdaExpr<T> = (T) -> T

val identity: LambdaExpr<Int> = { x -> x }
println(identity(5)) // Output: 5

// Step 2: Implement Church Booleans
typealias ChurchBoolean = <T> (T, T) -> T

val lcTrue: ChurchBoolean = { t, _ -> t }
val lcFalse: ChurchBoolean = { _, f -> f }
val lcIf: (ChurchBoolean, () -> Any, () -> Any) -> Any = { condition, then, else -> condition(then(), else()) }

println(lcIf(lcTrue, { "yes" }, { "no" })) // Output: yes

// Step 3: Implement Basic Combinators
val I: LambdaExpr<Any> = { x -> x }
val K: (Any) -> (Any) -> Any = { x -> { _ -> x } }
val KI: (Any) -> (Any) -> Any = { _ -> { y -> y } }

// Step 4: Create Church Numerals
typealias ChurchNumeral = (LambdaExpr<Int>) -> LambdaExpr<Int>

val lcZero: ChurchNumeral = { _ -> { x -> x } }
val lcSucc: (ChurchNumeral) -> ChurchNumeral = { n -> { f -> { x -> f(n(f)(x)) } } }

fun churchToInt(n: ChurchNumeral): Int = n({ it + 1 })(0)

println(churchToInt(lcSucc(lcZero))) // Output: 1

// Step 5: Implement Basic Arithmetic
val lcAdd: (ChurchNumeral) -> (ChurchNumeral) -> ChurchNumeral = { m -> { n -> { f -> { x -> m(f)(n(f)(x)) } } } }
val lcMult: (ChurchNumeral) -> (ChurchNumeral) -> ChurchNumeral = { m -> { n -> { f -> m(n(f)) } } }

val lcOne = lcSucc(lcZero)
val lcTwo = lcSucc(lcOne)

println(churchToInt(lcAdd(lcOne)(lcTwo))) // Output: 3

// Step 6: Implement Pairs
typealias Pair<T> = ((T, T) -> T) -> T

fun <T> lcPair(x: T, y: T): Pair<T> = { f -> f(x, y) }
fun <T> lcFirst(p: Pair<T>): T = p { x, _ -> x }
fun <T> lcSecond(p: Pair<T>): T = p { _, y -> y }

val myPair = lcPair("hello", "world")
println(lcFirst(myPair)) // Output: hello

// Step 7: Build Lists Using Pairs
typealias LCList<T> = Pair<T?>

val lcNil: LCList<Any?> = lcPair(null, null)
fun <T> lcCons(head: T, tail: LCList<T>): LCList<T> = lcPair(head, tail)
fun <T> lcIsNil(list: LCList<T>): Boolean = lcFirst(list) == null
fun <T> lcHead(list: LCList<T>): T? = lcFirst(list)
fun <T> lcTail(list: LCList<T>): LCList<T> = lcSecond(list) as LCList<T>

// Step 8: Implement Comparison Operations
val lcIsZero: (ChurchNumeral) -> ChurchBoolean = { n -> n({ _ -> lcFalse })(lcTrue) }
val lcLessThan: (ChurchNumeral) -> (ChurchNumeral) -> ChurchBoolean = { m -> { n ->
    n({ _ -> lcFalse })({ f ->
        m({ _ -> lcTrue })(f)
    })
} }
val lcEqual: (ChurchNumeral) -> (ChurchNumeral) -> ChurchBoolean = { m -> { n ->
    lcAnd(lcLessThan(m)(n))(lcLessThan(n)(m))
} }

// Helper functions for boolean operations
val lcAnd: (ChurchBoolean) -> (ChurchBoolean) -> ChurchBoolean = { p -> { q -> p(q)(p) } }
val lcOr: (ChurchBoolean) -> (ChurchBoolean) -> ChurchBoolean = { p -> { q -> p(p)(q) } }
val lcNot: (ChurchBoolean) -> ChurchBoolean = { p -> p(lcFalse)(lcTrue) }

// Step 9: Implement Higher-Order List Operations
fun <T, R> lcMap(f: (T) -> R, list: LCList<T>): LCList<R> =
    if (lcIsNil(list)) lcNil
    else lcCons(f(lcHead(list)!!), lcMap(f, lcTail(list)))

fun <T> lcFilter(predicate: (T) -> Boolean, list: LCList<T>): LCList<T> =
    if (lcIsNil(list)) lcNil
    else if (predicate(lcHead(list)!!)) lcCons(lcHead(list)!!, lcFilter(predicate, lcTail(list)))
    else lcFilter(predicate, lcTail(list))

fun <T, R> lcFold(f: (T, R) -> R, initial: R, list: LCList<T>): R =
    if (lcIsNil(list)) initial
    else lcFold(f, f(lcHead(list)!!, initial), lcTail(list))

// Step 10: Explore Recursion and Fixed-Point Combinators
typealias Recursive<T> = ((T) -> T) -> (T) -> T

val Y: Recursive<(Int) -> Int> = { f ->
    { x -> f({ y -> x(x)(y) })(x) }({ x -> f({ y -> x(x)(y) }) })
}

// Example: Factorial using Y combinator
val factorial = Y { f ->
    { n ->
        if (n <= 1) 1
        else n * f(n - 1)
    }
}

println(factorial(5)) // Output: 120

// Usage examples
fun main() {
    // Church numerals
    val lcThree = lcSucc(lcTwo)
    println("3 as Church numeral: ${churchToInt(lcThree)}")

    // List operations
    val list = lcCons(1, lcCons(2, lcCons(3, lcNil)))
    val doubled = lcMap({ it * 2 }, list)
    println("Doubled list: ${lcToString(doubled)}")

    val evens = lcFilter({ it % 2 == 0 }, list)
    println("Even numbers: ${lcToString(evens)}")

    val sum = lcFold({ a, b -> a + b }, 0, list)
    println("Sum of list: $sum")

    // Factorial
    println("Factorial of 5: ${factorial(5)}")
}

// Helper function to convert LCList to String for printing
fun <T> lcToString(list: LCList<T>): String {
    fun helper(l: LCList<T>, acc: List<T>): List<T> =
        if (lcIsNil(l)) acc
        else helper(lcTail(l), acc + lcHead(l)!!)
    return helper(list, emptyList()).toString()
}

File: TypeScript/lambdaCalculus.ts
/**
 * Lambda Calculus Implementation in TypeScript
 * Created by Claude 3.5 Sonnet (Anthropic, PBC)
 * 
 * Recommended filename: lambdaCalculus.ts
 * 
 * Usage:
 * 1. Ensure you have TypeScript installed
 * 2. Save this file as lambdaCalculus.ts
 * 3. Compile: tsc lambdaCalculus.ts
 * 4. Run: node lambdaCalculus.js
 */

// Step 1: Understand Lambda Expressions
type LambdaFunction<T, U> = (x: T) => U;

const identity: LambdaFunction<any, any> = (x) => x;
console.log(identity(5)); // Outputs: 5

// Step 2: Implement Church Booleans
type ChurchBoolean = <T>(x: T) => (y: T) => T;

const lcTrue: ChurchBoolean = (x) => (y) => x;
const lcFalse: ChurchBoolean = (x) => (y) => y;
const lcIf = <T>(condition: ChurchBoolean) => (then: T) => (else_: T): T => condition(then)(else_);

console.log(lcIf(lcTrue)('yes')('no')); // Outputs: 'yes'

// Step 3: Implement Basic Combinators
const I = identity;
const K: <T, U>(x: T) => (y: U) => T = (x) => (y) => x;
const KI: <T, U>(x: T) => (y: U) => U = (x) => (y) => y;

// Step 4: Create Church Numerals
type ChurchNumeral = <T>(f: (x: T) => T) => (x: T) => T;

const lcZero: ChurchNumeral = (f) => (x) => x;
const lcSucc = (n: ChurchNumeral): ChurchNumeral => (f) => (x) => f(n(f)(x));

const churchToInt = (n: ChurchNumeral): number => n((x: number) => x + 1)(0);

console.log(churchToInt(lcSucc(lcZero))); // Outputs: 1

// Step 5: Implement Basic Arithmetic
const lcAdd = (m: ChurchNumeral) => (n: ChurchNumeral): ChurchNumeral => (f) => (x) => m(f)(n(f)(x));
const lcMult = (m: ChurchNumeral) => (n: ChurchNumeral): ChurchNumeral => (f) => m(n(f));

const lcOne = lcSucc(lcZero);
const lcTwo = lcSucc(lcOne);

console.log(churchToInt(lcAdd(lcOne)(lcTwo))); // Outputs: 3

// Step 6: Implement Pairs
type ChurchPair<T, U> = <R>(f: (x: T) => (y: U) => R) => R;

const lcPair = <T, U>(x: T) => (y: U): ChurchPair<T, U> => (f) => f(x)(y);
const lcFirst = <T, U>(p: ChurchPair<T, U>): T => p((x) => (y) => x);
const lcSecond = <T, U>(p: ChurchPair<T, U>): U => p((x) => (y) => y);

const myPair = lcPair('hello')('world');
console.log(lcFirst(myPair)); // Outputs: 'hello'

// Step 7: Build Lists Using Pairs
type ChurchList<T> = ChurchPair<T, ChurchList<T>> | typeof lcNil;

const lcNil = lcPair(lcTrue)(lcTrue);
const lcCons = <T>(head: T) => (tail: ChurchList<T>): ChurchList<T> => lcPair(lcFalse)(lcPair(head)(tail));
const lcIsNil = <T>(list: ChurchList<T>): boolean => lcFirst(list as ChurchPair<ChurchBoolean, any>) === lcTrue;
const lcHead = <T>(list: ChurchList<T>): T => lcFirst(lcSecond(list as ChurchPair<ChurchBoolean, ChurchPair<T, any>>));
const lcTail = <T>(list: ChurchList<T>): ChurchList<T> => lcSecond(lcSecond(list as ChurchPair<ChurchBoolean, ChurchPair<T, ChurchList<T>>>));

// Step 8: Implement Comparison Operations
const lcIsZero = (n: ChurchNumeral): ChurchBoolean => n((x) => lcFalse)(lcTrue);
const lcLessThan = (m: ChurchNumeral) => (n: ChurchNumeral): ChurchBoolean => 
  lcNot(lcIsZero(lcSub(n)(m)));
const lcEqual = (m: ChurchNumeral) => (n: ChurchNumeral): ChurchBoolean => 
  lcAnd(lcLessThan(m)(lcSucc(n)))(lcLessThan(n)(lcSucc(m)));

// Helper functions for comparison operations
const lcNot = (b: ChurchBoolean): ChurchBoolean => (x) => (y) => b(y)(x);
const lcAnd = (a: ChurchBoolean) => (b: ChurchBoolean): ChurchBoolean => (x) => (y) => a(b(x)(y))(y);
const lcSub = (m: ChurchNumeral) => (n: ChurchNumeral): ChurchNumeral => 
  n((f) => (g) => (h) => h(f(g)))((f) => (x) => x)(m);

// Step 9: Implement Higher-Order List Operations
const lcMap = <T, U>(f: (x: T) => U) => (list: ChurchList<T>): ChurchList<U> =>
  lcIsNil(list) ? lcNil : lcCons(f(lcHead(list)))(lcMap(f)(lcTail(list)));

const lcFilter = <T>(pred: (x: T) => ChurchBoolean) => (list: ChurchList<T>): ChurchList<T> =>
  lcIsNil(list) ? lcNil :
    pred(lcHead(list)) ? lcCons(lcHead(list))(lcFilter(pred)(lcTail(list))) :
      lcFilter(pred)(lcTail(list));

const lcFold = <T, U>(f: (acc: U) => (x: T) => U) => (acc: U) => (list: ChurchList<T>): U =>
  lcIsNil(list) ? acc : lcFold(f)(f(acc)(lcHead(list)))(lcTail(list));

// Step 10: Explore Recursion and Fixed-Point Combinators
const Y = <T>(f: (g: (x: T) => T) => (x: T) => T): (x: T) => T =>
  ((x) => f((y) => x(x)(y)))((x) => f((y) => x(x)(y)));

// Example: Factorial using Y combinator
const factorial = Y((f) => (n: ChurchNumeral) =>
  lcIf(lcIsZero(n))
    (lcOne)
    (lcMult(n)(f(lcSub(n)(lcOne))))
);

console.log(churchToInt(factorial(lcSucc(lcSucc(lcSucc(lcZero)))))); // Outputs: 6 (3!)

// Helper function to convert regular numbers to Church numerals
const intToChurch = (n: number): ChurchNumeral => 
  n === 0 ? lcZero : lcSucc(intToChurch(n - 1));

// Example usage of higher-order list operations
const exampleList = lcCons(1)(lcCons(2)(lcCons(3)(lcNil)));
const doubledList = lcMap((x: number) => x * 2)(exampleList);
const evenNumbers = lcFilter((x: number) => lcEqual(intToChurch(x % 2))(lcZero))(exampleList);
const sum = lcFold((acc: number) => (x: number) => acc + x)(0)(exampleList);

console.log("Doubled list:", churchToInt(lcHead(doubledList) as ChurchNumeral)); // Outputs: 2
console.log("Even numbers:", churchToInt(lcHead(evenNumbers) as ChurchNumeral)); // Outputs: 2
console.log("Sum:", sum); // Outputs: 6
